src/sentry/uptime/endpoints/validators.py

```diff
@@ -463,44 +463,90 @@ def create_source(self, validated_data: dict[str, Any]) -> UptimeSubscription:
-            return config
+        request = self.context["request"]
+        is_superuser = is_active_superuser(request)
 
-        mode = config["mode"]
+        # On partial updates, merge attrs with existing config to preserve unspecified fields
+        # DRF's partial=True makes fields optional but doesn't auto-merge with instance data
+        if self.instance and self.partial:
+            existing_config = self.instance.copy()
+            existing_config.update(attrs)
+            attrs = existing_config
 
-        # For updates, if current mode is not MANUAL and user is not a superuser,
-        # automatically switch to MANUAL mode
+        # On updates, check if we need to auto-switch from AUTO_DETECTED to MANUAL
         if self.instance:
-            current_mode = self.instance.config.get("mode")
-
-            if current_mode != UptimeMonitorMode.MANUAL:
-                request = self.context["request"]
-                if not is_active_superuser(request):
-                    config["mode"] = UptimeMonitorMode.MANUAL
-                    return config
+            current_mode = self.instance.get("mode")
+            mode_in_request = "mode" in attrs
+            requested_mode = attrs.get("mode") if mode_in_request else current_mode
+
+            # If currently AUTO_DETECTED and not a superuser, force switch to MANUAL
+            if current_mode != UptimeMonitorMode.MANUAL and not is_superuser:
+                attrs["mode"] = UptimeMonitorMode.MANUAL
+
+            # If non-superuser is trying to change mode to something other than MANUAL
+            elif (
+                mode_in_request
+                and requested_mode != current_mode
+                and requested_mode != UptimeMonitorMode.MANUAL
+                and not is_superuser
+            ):
+                raise serializers.ValidationError({"mode": ["Only superusers can modify `mode`"]})
+        else:
+            # On create, non-superusers can only set MANUAL mode
+            if "mode" in attrs and attrs["mode"] != UptimeMonitorMode.MANUAL and not is_superuser:
+                raise serializers.ValidationError({"mode": ["Only superusers can modify `mode`"]})
 
-            # If mode hasn't changed, no further validation needed
-            if current_mode == mode:
-                return config
+        return attrs
 
-        # Only superusers can set/change mode to anything other than MANUAL
-        if mode != UptimeMonitorMode.MANUAL:
-            request = self.context["request"]
-            if not is_active_superuser(request):
-                raise serializers.ValidationError("Only superusers can modify `mode`")
 
-        return config
+class UptimeDomainCheckFailureValidator(BaseDetectorTypeValidator):
+    enforce_single_datasource = True
+    data_sources = serializers.ListField(child=UptimeMonitorDataSourceValidator(), required=False)
+    config = UptimeDomainCheckFailureConfigValidator(required=False)  # type: ignore[assignment]
```

Comment:
any thoughts on how we might be able to have a similar abstraction enforced through the base detector? my guess is that most detectors will have the not great error handling for the json config field.

it'd also be cool if we updated the base detector to allow for this kind of override without having to add a lint exception. 🤔

---

anywho, no action needed for this pr, just curious to get your take.